validator/lib/validateCmpt.js

```diff
@@ -0,0 +1,136 @@
+    }
+
+    var byteOffset = 16;
+    var sizeOfUint32 = 4;
+    var isValid = true;
+
+    var magic = content.toString('utf8', 0, 4);
+    var version = content.readUInt32LE(4);
+    var byteLength = content.readUInt32LE(8);
+    var tilesLength = content.readUInt32LE(12);
+
+    if (magic !== 'cmpt') {
+        return {
+            result : false,
+            message: 'Cmpt header has an invalid magic field. Expected magic = \'cmpt\'. Found magic = ' + magic
+        };
+    }
+
+    if (version !== 1) {
+        return {
+            result : false,
+            message: 'Cmpt header has an invalid version field. Expected version = 1. Found version = ' + version
+        };
+    }
+
+    if (byteLength !== content.length) {
+        return {
+            result : false,
+            message: 'Cmpt header has an invalid byteLength field. Expected byteLength = ' + content.length + '. Found byteLength = ' + byteLength
+        };
+    }
+
+    for (var i = 0; i < tilesLength; i++) {
+        var errorAddon = '';
+        var innerTileMagic = content.toString('utf8', byteOffset, byteOffset + sizeOfUint32);
+        var innerTileByteLength = content.readUInt32LE(byteOffset + 2 * sizeOfUint32);
+
+        var innerTile;
+        var validatorResult;
+
+        if (innerTileMagic === 'b3dm') {
+            innerTile = new Buffer(24);
+            content.copy(innerTile, 0, byteOffset, byteOffset + innerTileByteLength);
+            validatorResult = validateB3dm(innerTile);
+            isValid = isValid && validatorResult.result;
+        }
+
+        else if (innerTileMagic === 'i3dm'){
+            innerTile = new Buffer(32);
+            content.copy(innerTile, 0, byteOffset, byteOffset + innerTileByteLength);
+            validatorResult = validateI3dm(innerTile);
+            isValid = isValid && validatorResult.result;
+        }
+
+        else if (innerTileMagic === 'pnts'){
+            innerTile = new Buffer(28);
+            content.copy(innerTile, 0, byteOffset, byteOffset + innerTileByteLength);
+            validatorResult = validatePnts(innerTile);
+            isValid = isValid && validatorResult.result;
+        }
+
+        else if (innerTileMagic === 'cmpt') {
+            innerTile = new Buffer(innerTileByteLength);
+            content.copy(innerTile, 0, byteOffset, byteOffset + innerTileByteLength);
+            validatorResult = validateCmpt(innerTile);
+            isValid = isValid && validatorResult.result;
+        }
+
+        else {
+            isValid = false;
+            errorAddon += 'Inner tile header magic cannot be identified';
+        }
+
+        byteOffset = byteOffset + innerTileByteLength; // skip over this tile
+
+        if( (byteOffset + 12 > byteLength) && (i + 1 !== tilesLength) ) {
+            isValid = false;
+            errorAddon += 'Cmpt header given tilesLength = ' + tilesLength + '. Found number of inner tiles = ' + (i + 1);
+        }
```

Comment:
This check and the one at the end


```javascript
    if (byteOffset !== byteLength) {
        return {
            result : false,
            message: 'Header has invalid inner tile formats or tileLength field. Expected end of tile = ' + byteLength + '. Found end of tile = ' + byteOffset
        };
    }
```

can be combined together. And a different test can be added to differentiate between the tile byte length going too far vs. an inner tile not existing.

It will be clearer if all this is near the beginning of the loop. The message can also be simpler, just that an inner tile's byte length exceeds the cmpt tile's byte length. Before this check you can have another check that sees if `byteOffset + 3 * sizeOfUint32` exceeds `byteLength` (aka the buffer doesn't actually contain this tile and can't even get the tile's byte length). If this is true than the error message can be the one you have here, that the cmpt tile does not have as many inner tiles as it says it has.